### PR TITLE
Fix binding null refs

### LINKED ISSUE #388: ListBindingHelper.GetListItemType(typeof(IListSourceThatReturnsNull)) throws NRE should return null
Should return null

```cs
void Method()
{
      ListBindingHelper.GetListItemType(ilistSource.GetType());
}

        private class ClassWithIListSource : IListSource
        {
            public bool ContainsListCollection => throw new NotImplementedException();

            public IList GetListResult { get; set; }

            public IList GetList() => GetListResult;
        }
```

### LINKED ISSUE #389: ListBindingHelper.GetListItemType(new IEnumerableWrapper(null)) throws NRE should return typeof(object)
```cs
ListBindingHelper.GetListItemType(new IEnumerableWrapper(null));

private class IEnumerableWrapper : IEnumerable
{
    public IEnumerableWrapper(IList list)
    {
        List = list;
    }

    private IList List { get; }

    public IEnumerator GetEnumerator() => List?.GetEnumerator();
}
```

### LINKED ISSUE #390: ListBindingHelper.GetListItemProperties(x, new PropertyDescriptor[] { null }) throws NRE should return empty

### LINKED ISSUE #391: ListBindingHelper.GetListName(typeof(Array), new PropertyDescriptor[0]) throws NRE should return "Array"

### LINKED ISSUE #392: ListBindingHelper.GetListName(1, new PropertyDescriptor[] { null }) throws NRE should return null

### PR DESCRIPTION
**Pull request by hughbe**
Fixes #388 
Fixes #389
Fixes #390
Fixes #391
Fixes #392

This is an obvious win - we now return something meaningful (and usable) in these examples rather than throwing NRE. There are no changes in exceptions, only exceptions in edge cases that are no longer an issue!

### CODE REVIEW

**Comment by hughbe** [comment]
Tests with 100% coverage coming in separate PR

File: src/System.Windows.Forms/src/System/Windows/Forms/ListBindingHelper.cs [hunk 1]
```diff
@@ -442,7 +462,13 @@ public static class ListBindingHelper {
             PropertyDescriptorCollection pdc;
 
             // Get the value of the first listAccessor
-            if (listAccessors != null && listAccessors.Length > startIndex) {
+            if (listAccessors.Length > startIndex)
+            {
+                if (listAccessors[startIndex] == null)
+                {
+                    return new PropertyDescriptorCollection(null);
```
Comment (JuditRose) [inline src/System.Windows.Forms/src/System/Windows/Forms/ListBindingHelper.cs]: Wouldn't it make sense to initialize pdc as `new PropertyDescriptorCollection(null)` and just update the if statement above to `if (listAccessors != null && listAccessors.Length > startIndex && listAccessors[startIndex] != null)`? 
Removing `listAccessors != null` from the `if` wouldn't cause possible null reference exceptions?

Also these methods are written following the idea of keeping only one return statement instead of multiple returns, so I'd prefer to keep it that way and not mix it up if possible. (GetListItemPropertiesByInstance and GetListItemPropertiesByType both) What do you think?

File: src/System.Windows.Forms/src/System/Windows/Forms/ListBindingHelper.cs [hunk 2]
```diff
@@ -217,6 +218,11 @@ public static class ListBindingHelper {
             }
 
             list = GetList(list);
+            if (list == null)
```
Comment (JuditRose) [inline src/System.Windows.Forms/src/System/Windows/Forms/ListBindingHelper.cs]: These lines are duplication of L209-211, is it possible to move things around to get rid of the duplicity?

**Comment by JuditRose** [comment]
Besides minor things I noticed and mentioned above, LGTM.

**Comment by hughbe** [comment]
Tests are in this PR https://github.com/dotnet/winforms/pull/395 (I have validated that they pass with this change)
Fixed merge conflict

**Comment by zsd4yr** [comment]
I'll allow the separate PR for the tests, but let's try our best not to do this too much 😄
